src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunction.java

```diff
@@ -0,0 +1,258 @@
+    }
+    return moduleFileGlobals;
+  }
+
+  private static class GetModuleFileResult {
+
+    byte[] moduleFileContents;
+    // `registry` can be null if this module has a non-registry override.
+    @Nullable Registry registry;
+  }
+
+  @Nullable
+  private Optional<GetModuleFileResult> getModuleFile(ModuleKey key, StarlarkOverrideApi override,
+      Environment env) throws ModuleFileFunctionException, InterruptedException {
+    if (override instanceof NonRegistryOverride) {
+      // TODO: The repo name is not necessarily the same as the module name here. Fix later.
+      String repoName = key.getName();
+      RepositoryDirectoryValue repoDir = (RepositoryDirectoryValue) env.getValue(
+          RepositoryDirectoryValue.key(RepositoryName.createFromValidStrippedName(repoName)));
+      if (repoDir == null) {
+        return null;
+      }
+      RootedPath moduleFilePath = RootedPath.toRootedPath(
+          Root.fromPath(repoDir.getPath()), PathFragment.create("MODULE.bazel"));
+      if (env.getValue(FileValue.key(moduleFilePath)) == null) {
+        return Optional.empty();
+      }
+      GetModuleFileResult result = new GetModuleFileResult();
+      result.moduleFileContents = readFile(moduleFilePath.asPath());
+      return Optional.of(result);
+    }
+
+    List<String> registries = Objects.requireNonNull(REGISTRIES.get(env));
+    if (override instanceof RegistryOverride) {
+      String overrideRegistry = ((RegistryOverride) override).getRegistry();
+      if (!overrideRegistry.isEmpty()) {
+        registries = ImmutableList.of(overrideRegistry);
+      }
+    } else if (override != null) {
+      throw errorf("unrecognized override type %s", override.getClass().getName());
+    }
+    ImmutableList.Builder<Registry> registryObjects = new ImmutableList.Builder<>();
+    for (String registryUrl : registries) {
+      try {
+        registryObjects.add(registryFactory.getRegistryWithUrl(
+            registryUrl.replace("%workspace%", workspaceRoot.getPathString())));
+      } catch (URISyntaxException e) {
+        throw new ModuleFileFunctionException(e);
+      }
+    }
+    try {
+      return getModuleFileFromRegistries(key, registryObjects.build(), env.getListener());
+    } catch (IOException e) {
+      throw new ModuleFileFunctionException(e);
+    }
+  }
+
+  private static byte[] readFile(Path path) throws ModuleFileFunctionException {
+    try {
+      return FileSystemUtils.readWithKnownFileSize(path, path.getFileSize());
+    } catch (IOException e) {
+      throw new ModuleFileFunctionException(e);
+    }
+  }
+
+  private static Optional<GetModuleFileResult> getModuleFileFromRegistries(ModuleKey key,
+      List<Registry> registries, ExtendedEventHandler eventHandler)
+      throws IOException, InterruptedException {
+    GetModuleFileResult result = new GetModuleFileResult();
+    for (Registry registry : registries) {
+      Optional<byte[]> moduleFile = registry.getModuleFile(key, eventHandler);
```

Comment:
Can this call result in reading files? If so, the Skyframe dependencies on these files are missing (I know that because you are not passing in an `Environment`)